Increase unit test code coverage of `app/server` by 2.9%

AniDashyan opened the pull request:
#### Summary

This PR adds unit tests for `CommissioningWindowManager.cpp` in `app/server`, focusing on the `OnPlatformEvent` function. The tests improve coverage from 50.3% to 52.9% and validate scenarios such as:

* Successful commissioning completion
* Fail-safe timer expiration
* Eviction of an open PASE session
* Operational network enabled (success/failure)
* Stopping BLE advertisement after all BLE connections are closed

#### Implementation Notes
To improve testability and remove singleton dependencies, this PR introduces **dependency injection (DI**) for the DNS-SD server via `Server`:
* Added `DnssdServer` interface in `src/lib/dnssd` to abstract DNS-SD operations and reduce coupling.
* Introduced `PlatformDnssdServer` in `src/app/server` to bridge the `DnssdServer` interface with the existing `DnssdServer`.
* `Server::Init()` initializes a default `PlatformDnssdServer`.

#### Related issues

Main Issue [#37226 ](https://github.com/project-chip/connectedhomeip/issues/37226)

#### Testing

* Added `MockDnssdServer` for isolated testing of  `OnPlatformEvent`.
* Injected via `SetDnssdServer(&mMockDnssdServer)`  in `SetUpTestSuite/TearDownTestSuite`.
* Tests validate success/failure paths and stopped-server behavior.

AniDashyan committed:
Add unit tests for `OnPlarformEvent` when Commissioning is Completed

AniDashyan committed:
Add accessor class for `CommissioningWindowManager`

AniDashyan committed:
Fix setting the `mPASESession` properly

AniDashyan committed:
Add unit tests for  when allbleconnectionsareclosed, operationalnetwork is enabled

AniDashyan committed:
Improve  unit tests for `CommissioningWindowManager` to verify commissioning window state changes on platform events

AniDashyan committed:
Add helpful comments

AniDashyan committed:
Fix restyle issues

AniDashyan committed:
Fix restyle issues

AniDashyan committed:
Adds test for PASE session eviction on failsafe timer expiry

Adds a test case to verify that the PASE session is correctly evicted when the failsafe timer expires.
This ensures that resources are properly cleaned up and prevents potential security vulnerabilities.

Modifies the test access class to provide access to the PASE session.

AniDashyan committed:
Add helpful comments

AniDashyan committed:
Address gemini comments

Disable commissioning mode after PASE Session is evicted when failsafe timer expires

Fix typos and adjust comments

AniDashyan commented on src/app/tests/CommissioningWindowManagerTestAccess.h:
```diff
@@ -0,0 +1,42 @@
+namespace Test {
+/**
+ * @brief Class acts as an accessor to private methods of the CommissioningWindowManagerTestAccess class without needing to give
+ * friend access to each individual test.
```
Done

AniDashyan commented on src/app/tests/TestCommissioningWindowManager.cpp:
```diff
@@ -406,4 +409,119 @@ TEST_F(TestCommissioningWindowManager, TestCheckCommissioningWindowManagerEnhanc
+
+    commissionMgr.OnPlatformEvent(&event);
+    // Verify that PASE Session was evicted after failsafe timer expiration
+    EXPECT_FALSE(PASESession.Get().HasValue());
```
`MarkForEviction()` by itself doesn’t clear the `SessionHolder`, so checking `HasValue()` right after would normally be flaky.
In this code path, however, after the eviction check, `HandleFailedAttempt()` is invoked, which calls `Release()` on the `SessionHolder`. That guarantees that `HasValue()` will be false immediately.
Using `GetState()` isn’t possible here because the session has already been released, and dereferencing leads to a segfault. That’s why `EXPECT_FALSE(PASESession.Get().HasValue())` is the correct assertion.

AniDashyan committed:
Add a helper function for event creation

AniDashyan commented:
/gemini review

AniDashyan committed:
Resolve Gemini comments

AniDashyan commented on src/app/tests/TestCommissioningWindowManager.cpp:
```diff
@@ -406,4 +409,121 @@ TEST_F(TestCommissioningWindowManager, TestCheckCommissioningWindowManagerEnhanc
+}
+
+// Verify that BLE advertising is stopped when all BLE connections are closed
+TEST_F(TestCommissioningWindowManager, TestOnPlatformEventCloseCloseAllBleConnections)
```
Done.

AniDashyan commented on src/app/tests/TestCommissioningWindowManager.cpp:
```diff
@@ -406,4 +409,121 @@ TEST_F(TestCommissioningWindowManager, TestCheckCommissioningWindowManagerEnhanc
+    uint16_t peerSessionId                   = 2;
+    chip::FabricIndex fabricIndex            = chip::kUndefinedFabricIndex;
+    chip::Transport::PeerAddress peerAddress = chip::Transport::PeerAddress::UDP(
+        chip::Inet::IPAddress::Loopback(chip::Inet::IPAddressType::kAny), 5540); // 5540 is a port number assigned to Matter by IANA
```
Thanks. Done.

AniDashyan committed:
Fix restyle issues

AniDashyan committed:
Add an explanatory comment why the commissioning window is closed when commissioning is completed

AniDashyan committed:
Fix restyle issues

cecille commented on src/app/tests/CommissioningWindowManagerTestAccess.h:
```diff
@@ -0,0 +1,42 @@
+/*
+ *
+ *    Copyright (c) 2024 Project CHIP Authors
```
```suggestion
 *    Copyright (c) 2025 Project CHIP Authors
```

cecille commented on src/app/tests/TestCommissioningWindowManager.cpp:
```diff
@@ -406,4 +409,128 @@ TEST_F(TestCommissioningWindowManager, TestCheckCommissioningWindowManagerEnhanc
+
+// When BLE is enabled
+#if CONFIG_NETWORK_LAYER_BLE && CHIP_DEVICE_CONFIG_SUPPORTS_CONCURRENT_CONNECTION
+    EXPECT_FALSE(Server::GetInstance().GetBleLayerObject()->IsBleClosing());
```
Have you tried this test with these on? I don't think this will work if you've started advertisements with dnssd only.

AniDashyan replied:
Both `CONFIG_NETWORK_LAYER_BLE` and `CHIP_DEVICE_CONFIG_SUPPORTS_CONCURRENT_CONNECTION` are enabled by default, so this block is compiled and entered during tests. I’ve been running with BLE enabled, and the test passes in that mode. You’re correct that if only DNS-SD advertisements are started the behavior may differ. I’ll re-run the test with the commissioning window set to `kAllSupported` to ensure BLE is active and confirm the expectation is valid in that case as well.

cecille commented on src/app/tests/TestCommissioningWindowManager.cpp:
```diff
@@ -406,4 +409,128 @@ TEST_F(TestCommissioningWindowManager, TestCheckCommissioningWindowManagerEnhanc
+    commissionMgr.OnPlatformEvent(&event);
+    // When the commissioning has completed (kCommissioningComplete event) OnPlatformEvent cleans up, closes active sessions and the
+    // commissioning window is closed The device should no longer be discoverable or accept new commissioners
+    EXPECT_FALSE(commissionMgr.IsCommissioningWindowOpen());
```
Is it also possible to verify the advertisement state for mDNS?

cecille commented on src/app/tests/TestCommissioningWindowManager.cpp:
```diff
@@ -406,4 +409,128 @@ TEST_F(TestCommissioningWindowManager, TestCheckCommissioningWindowManagerEnhanc
+    auto event = CreateEvent(chip::DeviceLayer::DeviceEventType::kFailSafeTimerExpired);
+
+    commissionMgr.OnPlatformEvent(&event);
+    EXPECT_TRUE(commissionMgr.IsCommissioningWindowOpen());
```
Wait, so the commissioning window manager doesn't auto-close the commissioning window on failsafe expiry?

cecille commented on src/app/tests/TestCommissioningWindowManager.cpp:
```diff
@@ -406,4 +409,128 @@ TEST_F(TestCommissioningWindowManager, TestCheckCommissioningWindowManagerEnhanc
+#endif
+}
+
+// Verify that the when fail-safe timer expires commissioning mode is disabled
```
This test isn't testing that. The testing is closing the window itself. What is the API supposed to look like here? Is this expected or a bug?

AniDashyan replied:
In the case of a normal failsafe timer expiry the commissioning window remains open is expected. The explicit `CloseCommissioningWindow()` call is included for cleanup to ensure subsequent tests run safely and it also verifies that the commissioning mode changes to `kDisabled` after closure.

cecille commented on src/app/tests/TestCommissioningWindowManager.cpp:
```diff
@@ -406,4 +409,128 @@ TEST_F(TestCommissioningWindowManager, TestCheckCommissioningWindowManagerEnhanc
+    auto event = CreateEvent(chip::DeviceLayer::DeviceEventType::kOperationalNetworkEnabled);
+
+    commissionMgr.OnPlatformEvent(&event);
+    EXPECT_EQ(chip::app::DnssdServer::Instance().AdvertiseOperational(), CHIP_NO_ERROR);
```
This isn't checking that the commissioning manager started the advertisement - it's starting the advertisement.

cecille commented on src/app/tests/TestCommissioningWindowManager.cpp:
```diff
@@ -406,4 +409,128 @@ TEST_F(TestCommissioningWindowManager, TestCheckCommissioningWindowManagerEnhanc
+    auto event = CreateEvent(chip::DeviceLayer::DeviceEventType::kCloseAllBleConnections);
+
+    commissionMgr.OnPlatformEvent(&event);
+    EXPECT_FALSE(chip::DeviceLayer::ConnectivityMgr().IsBLEAdvertisingEnabled());
```
Do you know that it was started before the event was generated?

AniDashyan replied:
Before the event is generated, BLE advertising is not active. This happens because in the preceding test the server is stopped and restarted, which disables BLE advertising. Attempting to open a new commissioning window in this test causes a segfault. I’ll look into a safe way to test this behavior.

AniDashyan committed:
Update src/app/tests/CommissioningWindowManagerTestAccess.h

Co-authored-by: C Freeman <cecille@google.com>

AniDashyan committed:
Change advertising mode from `kDnssdOnly` to `kAllSupported` to enusre that when  BLE is enabled it closes properly.

Verify that when commissioning is completed successfully the commissioning mode is disabled.

AniDashyan committed:
Merge branch 'unit-tests-app-server' of github.com:AniDashyan/connectedhomeip into unit-tests-app-server

AniDashyan committed:
Remove redundant advertisement check

AniDashyan committed:
restore `FabricTable` after `StopServer()`  to avoid segfaults in subsequent tests

restyled-commits committed:
Restyled by whitespace

restyled-commits committed:
Restyled by clang-format

AniDashyan commented:
/gemini review

AniDashyan committed:
Fix a typo

AniDashyan committed:
Exclude `TestOnPlatformEventCloseAllBleConnections` test on Darwin since BLE advertisement is not supported

AniDashyan committed:
Add dependency injection for CommissioningManager to make it more testable

restyled-commits committed:
Restyled by whitespace

restyled-commits committed:
Restyled by clang-format

tcarmelveilleux commented on src/app/server/CommissioningWindowManager.h:
```diff
@@ -232,6 +237,11 @@ class CommissioningWindowManager : public Messaging::UnsolicitedMessageHandler,
+
+    // For testing purposes use a dependency injection
+    IDnssdServer *mDnsSdServer;
+    DnnsdServerWrapper mDefaultDnssd;
```
I would say it should be always possible to dependency-inject, likely from Server.cpp for now.

AniDashyan replied:
Thank you. Done.

AniDashyan commented on src/app/server/IDnssdServer.h:
```diff
@@ -0,0 +1,43 @@
+
+namespace chip  {
+    // namespace app {
+        class IDnssdServer
```
Do you suggest moving this file to `lib/dnssd`?

AniDashyan committed:
Rename class names in `IDnssdServer` for better clarity

AniDashyan commented on src/app/server/IDnssdServer.h:
```diff
@@ -0,0 +1,43 @@
+                virtual bool IsAdvertisingEnabled() = 0;
+        };
+
+        class DnnsdServerWrapper : public IDnssdServer 
```
Done.

restyled-commits committed:
Restyled by clang-format

AniDashyan committed:
Rename `chip::DnssdServer` to `chip::IDnssdServer` to avoid naming ambiguity

restyled-commits committed:
Restyled by clang-format

AniDashyan committed:
Remove `IDnssd.h`from server build configuration

AniDashyan committed:
Remove `IDnssd.h`from server build configuration

AniDashyan committed:
Merge branch 'unit-tests-app-server' of github.com:AniDashyan/connectedhomeip into unit-tests-app-server

AniDashyan committed:
Seperate the implementation of `PlatformDnssdServer` from `IDnssdServer` to avoid `#include <app/server/Dnssd.h>` errors in CI

Updates mock implementation in tests to align with the new interface
and refactors `IDnssdServer` to include `Start` and `Stop` methods.

AniDashyan committed:
Fix restyle issues

AniDashyan committed:
Merge branch 'master' into unit-tests-app-server

AniDashyan committed:
Refactor `CommissioningWindowManager`to use `Server`'s `DnssdServer` and update related methods

AniDashyan committed:
Merge branch 'unit-tests-app-server' of github.com:AniDashyan/connectedhomeip into unit-tests-app-server

restyled-commits committed:
Restyled by clang-format

AniDashyan commented:
/Gemini review

AniDashyan committed:
Set the `MockDnssdServer` for testing from `SetUpTestSuite` and reset to the real one in `TearDownTestSuite`

- Rename `Star`t and `Stop` methods to `StartServer` and `StopServer` for clarity

AniDashyan commented:
/gemini review

AniDashyan committed:
Remove redundant commented code

AniDashyan commented on src/app/tests/TestCommissioningWindowManager.cpp:
```diff
@@ -406,4 +443,145 @@ TEST_F(TestCommissioningWindowManager, TestCheckCommissioningWindowManagerEnhanc
+
+    commissionMgr.OnPlatformEvent(&event);
+    EXPECT_TRUE(mMockDnssd.IsAdvertisingEnabled());
+    // Server::GetInstance().SetDnssdServer(&(Server::GetInstance().GetDefaultDnssdServer()));
```
Thanks. Done.

AniDashyan commented on src/app/tests/TestCommissioningWindowManager.cpp:
```diff
@@ -406,4 +443,145 @@ TEST_F(TestCommissioningWindowManager, TestCheckCommissioningWindowManagerEnhanc
+{
+    // MockDnssdServer mockDnssd;
+    CommissioningWindowManager & commissionMgr = Server::GetInstance().GetCommissioningWindowManager();
+    // Server::GetInstance().SetDnssdServer(&mockDnssd);
```
Thanks. Done.

restyled-commits committed:
Restyled by clang-format

tcarmelveilleux commented on src/lib/dnssd/IDnssdServer.h:
```diff
@@ -0,0 +1,33 @@
+
+namespace chip {
+namespace Dnssd {
+class IDnssdServer
```
Please do not use `I` prefix. Please call this DnssdServer.

Also, please document all teh methods and what they do.

AniDashyan replied:
Thanks. Done.

tcarmelveilleux commented on src/app/tests/CommissioningWindowManagerTestAccess.h:
```diff
@@ -0,0 +1,42 @@
+{
+public:
+    CommissioningWindowManagerTestAccess() = delete;
+    CommissioningWindowManagerTestAccess(CommissioningWindowManager * cmwinMgr) : mCmwinMgr(cmwinMgr) {}
```
cmwinMgr --> please use a better name, it's OK to use full words.

tcarmelveilleux commented on src/app/tests/TestCommissioningWindowManager.cpp:
```diff
@@ -406,4 +443,145 @@ TEST_F(TestCommissioningWindowManager, TestCheckCommissioningWindowManagerEnhanc
+
+    commissionMgr.OnPlatformEvent(&event);
+    // This should attempt to start operational advertising, which will fail
+    EXPECT_EQ(mMockDnssd.AdvertiseOperational(), CHIP_ERROR_INCORRECT_STATE);
```
+1

AniDashyan replied:
Done.

tcarmelveilleux commented on src/app/server/Server.h:
```diff
@@ -467,9 +472,10 @@ class Server
     }
 
     static Server & GetInstance() { return sServer; }
+    void SetDnssdServer(Dnssd::IDnssdServer * dnssdServer) { mDnssdServer = dnssdServer; }
```
Please remove this setter.

tcarmelveilleux commented on src/app/server/CommissioningWindowManager.h:
```diff
@@ -30,8 +30,16 @@
 #include <protocols/secure_channel/PASESession.h>
 #include <system/SystemClock.h>
 
+#include <app/server/PlatformDnssdServer.h>
```
If this is what gets injected elsewhere, it's not needed as a dependency here. Please remove this dependency and only depend on DnssdServer.

tcarmelveilleux commented on src/app/server/PlatformDnssdServer.h:
```diff
@@ -0,0 +1,18 @@
+    CHIP_ERROR AdvertiseOperational() override { return DnssdServer::Instance().AdvertiseOperational(); }
+    virtual void StartServer() override { DnssdServer::Instance().StartServer(); }
+    virtual void StopServer() override { DnssdServer::Instance().StopServer(); }
+    bool IsAdvertisingEnabled() override
+    {
+        return false; // placeholder
+    }
```
Then why does the interface have it?

AniDashyan replied:
The interface has `IsAdvertisingEnabled()` method to check the advertising state, mainly for testing purposes. However, `chip::Dnssd::ServiceAdvertiser` doesn’t provide a public `IsAdvertising()` method, so the current implementation just returns `false` as a temporary placeholder.

AniDashyan committed:
Address comments

TODO: Refactor or remove the `TestOnPlatformEventFailSafeTimerExpiredPASEEVicted` test

AniDashyan committed:
Address comments

TODO: Refactor or remove the `TestOnPlatformEventFailSafeTimerExpiredPASEEVicted` test

AniDashyan committed:
Merge branch 'unit-tests-app-server' of github.com:AniDashyan/connectedhomeip into unit-tests-app-server

restyled-commits committed:
Restyled by whitespace

restyled-commits committed:
Restyled by clang-format

AniDashyan committed:
Improve `MockDnssdServer` class

AniDashyan committed:
Rename `IDnssdServer` to `DnssdServer` in `BUILD.gn`

AniDashyan committed:
Merge branch 'master' into unit-tests-app-server

AniDashyan committed:
Remove PASE eviction test

AniDashyan committed:
Remove PASE eviction test

AniDashyan committed:
Remove unused source file from test suite

AniDashyan committed:
Merge branch 'unit-tests-app-server' of github.com:AniDashyan/connectedhomeip into unit-tests-app-server

AniDashyan committed:
Merge branch 'unit-tests-app-server' of github.com:AniDashyan/connectedhomeip into unit-tests-app-server

AniDashyan committed:
Merge branch 'unit-tests-app-server' of github.com:AniDashyan/connectedhomeip into unit-tests-app-server

AniDashyan committed:
Merge branch 'master' into unit-tests-app-server

AniDashyan committed:
Merge branch 'master' into unit-tests-app-server

AniDashyan committed:
Fix: Set DNS-SD server on CommissioningWindowManager during init

AniDashyan committed:
Merge branch 'master' into unit-tests-app-server

AniDashyan commented:
/gemini review

AniDashyan commented on src/app/server/PlatformDnssdServer.h:
```diff
@@ -0,0 +1,18 @@
+    CHIP_ERROR AdvertiseOperational() override { return chip::app::DnssdServer::Instance().AdvertiseOperational(); }
+    void StartServer() override { chip::app::DnssdServer::Instance().StartServer(); }
+    void StopServer() override { chip::app::DnssdServer::Instance().StopServer(); }
+    bool IsAdvertisingEnabled() override
+    {
+        return false; // placeholder
+    }
```
`IsAdvertisingEnabled()` is only used in tests to verify advertising state. Since `chip::app::DnssdServer` doesn't expose this state, I'll add a comment explaining why it returns `false` as a placeholder.

AniDashyan replied:
՝IsAdvertisingEnabled()՝ is only used in tests. So instead of keeping it in the interface with a placeholder, I'll remove it from ՝DnssdServer՝ entirely and make it a test-specific method in ՝MockDnssdServer՝ only.

AniDashyan committed:
Address Gemini comments

AniDashyan committed:
Mark `PlatformDnssdServer` class as final and add inline specifiers to methods

AniDashyan committed:
Merge branch 'master' into unit-tests-app-server

AniDashyan committed:
Merge branch 'master' into unit-tests-app-server

AniDashyan committed:
Merge branch 'master' into unit-tests-app-server

LyudmilaKostanyan committed:
Merge branch 'master' into unit-tests-app-server
